File: 242.valid-anagram.rb
#
# @lc app=leetcode id=242 lang=ruby
#
# [242] Valid Anagram
#

# @lc code=start
# @param {String} s
# @param {String} t
# @return {Boolean}
def is_anagram(s, t)
  char_to_count = s.chars.tally

  t.chars.each do |c|
    return false unless char_to_count[c] && char_to_count[c] > 0

    char_to_count[c] -= 1
  end

  char_to_count.values.all? { |v| v == 0 }
end
# @lc code=end

puts(is_anagram('anagram', 'nagaram'))
puts(is_anagram('rat', 'car'))

File: 242.valid-anagram.go
/*
 * @lc app=leetcode id=242 lang=golang
 *
 * [242] Valid Anagram
 */
package main

// @lc code=start
func isAnagram(s string, t string) bool {
	exists := make(map[byte]int)

	for i := 0; i < len(s); i++ {
		exists[s[i]]++
	}

	for i := 0; i < len(t); i++ {
		exists[t[i]]--
	}

	for _, val := range exists {
		if val != 0 {
			return false
		}
	}

	return true
}

// @lc code=end
